### PR TITLE
Implement #162 - Verify trip usability and usage

### PR DESCRIPTION
**Pull request by lionel-nj**
**Summary:**

This PR provides support to verify that:
- all trips are referred to at least once in `stop_times.txt` (rule E050)
- all trip serve more than 1 stop (rule E051)

**Expected behavior:** 
Use cases E050 and E051 should be executed:

**E050:** if `trips.txt` defines a `trip_id` that is not used in `stop_times.txt` a `TripNotUsedNotice` is generated and added to the validation result repository

**E051:** if a trip does not serve more than one stop in `stop_times.txt` a `UnusableTripNotice` is generated and added to the validation result repository

Capture d’écran, le 2020-08-18 à 09 32 53

Please make sure these boxes are checked before submitting your pull request - thanks!

- [X] Run the unit tests with `gradle test` to make sure you didn't break anything
- [X] Format the title like "Fix #<issue_number> - <short description of fix and changes>" (for example - "Fix #1111 - Check for null value before using field")
- [X] Linked all relevant issues
- [X] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

File: usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateTripUsageTest.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import static org.mockito.Mockito.*;
+
+class ValidateTripUsageTest {
+
+    @Test
+    void usedTripShouldNotGenerateNotice() {
+        final Logger mockLogger = mock(Logger.class);
+        final GtfsDataRepository mockDataRepo = mock(GtfsDataRepository.class);
+        final Map<String, Trip> mockTripCollection = new HashMap<>();
+        mockTripCollection.put("trip id", mock(Trip.class));
```
Comment (barbeau) [inline usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateTripUsageTest.java:43]: I would suggest using something that looks like a real trip ID here, like `12345`, to make the code easier to read and distinguish from the field name `trip_id`.

File: usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateTripUsageTest.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        when(mockDataRepo.getTripAll()).thenReturn(mockTripCollection);
+
+        final Map<String, TreeMap<Integer, StopTime>> mockStopTimeCollection = new HashMap<>();
+        final TreeMap<Integer, StopTime> innerStopTimeMap = new TreeMap<>();
+        mockStopTimeCollection.put("trip id", innerStopTimeMap);
```
Comment (barbeau) [inline usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateTripUsageTest.java:48]: Same as above

File: usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateTripUsageTest.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        verifyNoMoreInteractions(mockDataRepo, mockLogger);
+        verifyNoInteractions(mockResultRepo);
+    }
+
+    @Test
+    void unusedTripShouldGenerateNotice() {
+        final Logger mockLogger = mock(Logger.class);
+        final GtfsDataRepository mockDataRepo = mock(GtfsDataRepository.class);
+        final Map<String, Trip> mockTripCollection = new HashMap<>();
+        mockTripCollection.put("unused trip id", mock(Trip.class));
```
Comment (barbeau) [inline usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateTripUsageTest.java:70]: Same as above

File: usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateTripUsageTest.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        when(mockDataRepo.getTripAll()).thenReturn(mockTripCollection);
+
+        final Map<String, TreeMap<Integer, StopTime>> mockStopTimeCollection = new HashMap<>();
+        final TreeMap<Integer, StopTime> innerStopTimeMap = new TreeMap<>();
+        mockStopTimeCollection.put("trip id", innerStopTimeMap);
```
Comment (barbeau) [inline usecase/src/test/java/org/mobilitydata/gtfsvalidator/usecase/ValidateTripUsageTest.java:75]: Same as above

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/UnusableTripNotice.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import org.mobilitydata.gtfsvalidator.domain.entity.notice.NoticeExporter;
+import org.mobilitydata.gtfsvalidator.domain.entity.notice.base.ErrorNotice;
+
+import java.io.IOException;
+
+public class UnusableTripNotice extends ErrorNotice {
+    public UnusableTripNotice(final String tripId) {
+        super("stop_times.txt", E_051,
+                "Trips must have more than one stop to be usable",
+                String.format("Trip with `trip_id`: `%s` should at least have two stops in stop_times.txt", tripId),
```
[the remark this answers is not in the document]
Comment (lionel-nj) [reply domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/UnusableTripNotice.java:28]: @LeoFrachet suggested to add ticks around the value we refer to improve the error report legibility and ease the task of data providers when it comes to fixing issues related to their datasets. 
For example, if we do not have these ticks and the error report claims an error for trip with trip_id `trip id 00  `, it can be difficult to track down this entity given the "invisible" (space) postfixing the "visible" characters. Although this particular situation is borderline, it can happen, and that is why I put the ticks around both the field name and value.

However, I'm totally open to change this if required, for consistency we should update all notice using the same pattern.

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/EmptyFileErrorNotice.java [hunk 6]
```diff
@@ -23,7 +23,7 @@
 
 public class EmptyFileErrorNotice extends ErrorNotice {
     public EmptyFileErrorNotice(String filename) {
-        super(filename, E_039,
+        super(filename, E_047,
```
[the remark this answers is not in the document]
Comment (lionel-nj) [reply domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/EmptyFileErrorNotice.java:26]: It appeared that this error code E039 was not the error code to be used for `EmptyFileErrorNotice` to ensure consistency with `RULES.md` doc. This probably comes an error when merging https://github.com/MobilityData/gtfs-validator/pull/317. 

Should I rather open a new issue referencing this, and apply this change in a new PR, so that this PR does not contain changes unrelated to  https://github.com/MobilityData/gtfs-validator/issues/162? @barbeau

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/UnusableTripNotice.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (barbeau) [reply domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/UnusableTripNotice.java:28]: That's fine, it can remain as-is. I would suggest getting feedback from providers at some point to make sure the assumption is correct.

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/EmptyFileErrorNotice.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (barbeau) [reply domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/EmptyFileErrorNotice.java:26]: Yes, I'd reference this in a new issue and separate PR, to avoid mixing unrelated things there.

[the remark this answers is not in the document]
Comment (lionel-nj) [reply domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/EmptyFileErrorNotice.java]: See https://github.com/MobilityData/gtfs-validator/issues/345

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/BackwardsTimeTravelInStopNotice.java [hunk 7]
```diff
@@ -27,7 +27,7 @@ public BackwardsTimeTravelInStopNotice(final String tripId,
                                            final String arrivalTimeAsString,
                                            final String previousStopDepartureTimeAsString,
                                            final Integer previousStopTimeStopSequence) {
-        super("stop_times.txt", E_047,
```
Comment (barbeau) [inline domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/BackwardsTimeTravelInStopNotice.java:30]: Same here in terms of an unrelated change - was this intended, a bug fix, or an bad merge?

Comment (lionel-nj) [reply domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/error/BackwardsTimeTravelInStopNotice.java:30]: Thanks, good catch I missed that one. It was intentional, so I'm removing this. Changes will be provided in a separate PR as previously discussed.
